### PR TITLE
CI: Run clang-tidy on the Threads backend

### PR DESCRIPTION
**Pull request by masterleinad**
This pull request tries to expand our `clang-tidy` coverage with the `Thread` backend.

### CODE REVIEW

File: .github/workflows/continuous-integration-workflow.yml [hunk 1]
```diff
@@ -13,24 +13,40 @@ jobs:
...
             cmake_build_type: 'Debug'
-            openmp: 'ON'
+            backend: 'OPENMP'
+            hwloc: 'ON'
+            clang-tidy: ''
+          - distro: 'ubuntu:latest'
+            cxx: 'clang++'
+            cmake_build_type: 'RelWithDebInfo'
+            backend: 'THREADS'
+            hwloc: 'OFF'
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply .github/workflows/continuous-integration-workflow.yml:48]: See #4878.

**Comment by masterleinad** [comment]
> Why are the hwloc changes part of this PR?

See https://github.com/kokkos/kokkos/pull/4842#pullrequestreview-911839100.

**Review by cz4rs — APPROVED** [review]
Looks good to me :+1:
The only change that makes me cautious is the `const_cast<int32_t volatile*>` in `Kokkos_TaskBase.hpp`, but as far as I can see this is correct as well.

File: .github/workflows/continuous-integration-workflow.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (MikolajZuzek) [reply .github/workflows/continuous-integration-workflow.yml:48]: Ok, thanks. If _Threads_ are worthy covering in CI, maybe it'd be good to add `g++` to the config too ?  
(_OpenMP_ builds for both `g++` and `clang++`, and also for `Debug` and `Release`).

**Review by Rombur — APPROVED** [review]

**Comment by dalg24** [comment]
CI build that was modified to run clang-tidy fails
```
/__w/kokkos/kokkos/core/src/impl/Kokkos_LIFO.hpp:80:36: error: do not use C-style cast to convert between unrelated types [cppcoreguidelines-pro-type-cstyle-cast,-warnings-as-errors]
[18](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:19)
  OwningRawPtr<node_type> m_head = (node_type*)EndTag;
[19](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:20)
                                   ^
[20](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:21)
/__w/kokkos/kokkos/core/src/impl/Kokkos_LIFO.hpp:92:24: error: do not use C-style cast to convert between unrelated types [cppcoreguidelines-pro-type-cstyle-cast,-warnings-as-errors]
[21](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:22)
    while (old_head != (node_type*)LockTag) {
[22](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:23)
                       ^
[23](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:24)
/__w/kokkos/kokkos/core/src/impl/Kokkos_LIFO.hpp:135:49: error: do not use C-style cast to convert between unrelated types [cppcoreguidelines-pro-type-cstyle-cast,-warnings-as-errors]
[24](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:25)
    return (volatile node_type*)this->m_head == (node_type*)EndTag;
[25](https://github.com/kokkos/kokkos/runs/6556204355?check_suite_focus=true#step:9:26)

**Comment by masterleinad** [comment]
`KokkosCore_UnitTest_CudaTimingBased`(`cuda.debug_pin_um_to_host`) failing for `CUDA-11.0-NVCC-C++17-RDC` looks unrelated.

**Comment by dalg24** [comment]
Rewrite history or squash?

**Comment by masterleinad** [comment]
Squashing is fine.

**Pull request merged**